core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static io.trino.operator.RetryPolicy.QUERY;
+import static io.trino.operator.RetryPolicy.TASK;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> finishedTasks = new HashSet<>();
+    @GuardedBy("this")
+    private final Map<TaskId, Throwable> failedTasks = new HashMap<>();
+    @GuardedBy("this")
+    private boolean inputFinished;
+    @GuardedBy("this")
+    private Throwable failure;
+
+    @GuardedBy("this")
+    private final ListMultimap<TaskId, SerializedPage> pageBuffer = ArrayListMultimap.create();
+    @GuardedBy("this")
+    private Iterator<SerializedPage> pagesIterator;
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+
+    @GuardedBy("this")
+    private boolean closed;
+
+    public DeduplicationExchangeClientBuffer(Executor executor, DataSize bufferCapacity, RetryPolicy retryPolicy)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+        this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    @Override
+    public synchronized SerializedPage pollPage()
+    {
+        throwIfFailed();
+
+        if (closed) {
+            return null;
+        }
+
+        if (!inputFinished) {
+            return null;
+        }
+
+        if (pagesIterator == null) {
+            pagesIterator = pageBuffer.values().iterator();
+        }
+
+        if (!pagesIterator.hasNext()) {
+            return null;
+        }
+
+        SerializedPage page = pagesIterator.next();
+        pagesIterator.remove();
```

Comment:
how costly is this one? Given you are using `ArrayListMultimap` it feels like it may be O(sth^2). Hard to figure out from the code though.

Reply:
Yeah, you are right, not sure what I was thinking :-) Let me switch it to `LinkedListMultimap` for now. It could be optimized further though. However I'm not sure if it makes sense to overspend time on this implementation as with adding spilling capabilities it may change.